IPs/RSKIP_rent2021.md

```diff
@@ -0,0 +1,187 @@
+
+
+## Rationale
+
+The specification already provides much of the rationale. More information to be added here based on community feedback.
+
+## Backwards Compatibility
+
+As already mentioned, this change needs a hard fork. In most cases, increasing the `gaslimit` should enough to account for rent payments. RPC method to `estimate_gas` will provide reasonable guidance. However, there can be some breaking changes.
+
+*Value-transferring* `CALL`s cost 9000 gas. From this 9000, an amount 2300 gas is subtracted as a *call stipend* and passed to the *receiving* contract. The stipend is intended to cover the cost for logging the value transfer. If the receiving contract performs any  `BAL` or `SLOAD` operations as part of logging, and if enough rent has accumulated in those accounts or storage cells, then some CALLs may fail.
+
+Contracts that make CALLs to external contacts with specific gas limits can also fail. Such patterns have been discouraged for quite some time now, at least since EIP-150, which increased awareness that gas costs can change  and developers should not make their execution logic rely on the gas schedule. 
+
+We think such breaks will be rare in RSK. We also note that the Ethereum community has deliberated and implemented potentially breaking changes of similar magnitude, most recently EIP-2929 which increased trie read operations by 3X.
+
+## Implementations
+
+Previously, the authors implemented a version with significant overlap with the current proposal. That implementation is being modified and will be shared with the community. This RSKIP will be modified with updated information at that point.
+
+### Implementation notes (RSKJ)
+
+Trie related
+- Use `long` for `lastRentPaidTime` with a default value of -1 (minus one). This value can help distinguish missing timestamps (which will default to 0). Recall that the trie data structure is also used for other tries (transactions, receipts)  
+- RSKJ currently uses an *implicit* node version number `01` as part of `flags` (bit positions 6, 7). For serialization, these bit positions should be changed to version 2 i.e. `10` in `flags`. 
+- Can use `nodeVersion` 0 when reading nodes with `Orchid` serialization format using `fromMessageOrchid()`. This will not affect encoding or hashing for Orchid.
+- the trie `put(key, value)`  method has to be *generalized* to `put(key, value, lastRentPaidTime)` to incorporate  `lastRentPaidTime`. The default value `put(key, value, -1` can be used for puts that do not need a timestamp (e.g. version 1 nodes, transaction tires, receipts tries)
+
+
+## Test Cases
+
+Incomplete
```

Comment:
Just in case, this Incomplete here is on purpose?

Reply:
Thanks! Yes, the "incomplete" is intentional. I need to specify the tests more clearly and add more of them.